src/freertos_drivers/esp32/Esp32WS2812.cxx

```diff
@@ -0,0 +1,311 @@
+/// @param item_num is the number of RMT items translated by this method.
+///
+/// Note: this method is called from an ISR context and all data access must be
+/// from DRAM, be a constant and be 32 bit aligned.
+static void IRAM_ATTR ws2812_rmt_adapter(const void *src, rmt_item32_t *dest,
+    size_t src_size, size_t wanted_num, size_t *translated_size,
+    size_t *item_num)
+{
+    if (src == NULL || dest == NULL)
+    {
+        *translated_size = 0;
+        *item_num = 0;
+        return;
+    }
+    size_t size = 0;
+    size_t num = 0;
+    uint8_t *psrc = (uint8_t *)src;
+    rmt_item32_t *pdest = dest;
+    const rmt_item32_t bit0 = {{{ timeZeroHighTicks, 1, timeZeroLowTicks, 0 }}};
+    const rmt_item32_t bit1 = {{{ timeOneHighTicks, 1, timeOneLowTicks, 0 }}};
+    while (size < src_size && num < wanted_num)
+    {
+        LOG(VERBOSE, "ESP-WS2812: %zu:%02x", size, *psrc);
+        for (int i = 7; i >= 0; i--)
+        {
+            // bytes are sent in MSB order
+            if (*psrc & (1 << i))
+            {
+                LOG(VERBOSE, "ESP-WS2812: %zu:%d 1", size, i);
+                pdest->val =  bit1.val;
+            }
+            else
+            {
+                LOG(VERBOSE, "ESP-WS2812: %zu:%d 0", size, i);
+                pdest->val =  bit0.val;
+            }
+            num++;
+            pdest++;
+        }
+        size++;
+        psrc++;
+    }
+
+    *translated_size = size;
+    *item_num = num;
+}
+
+// constructor
+Esp32WS2812::Esp32WS2812(const gpio_num_t output_pin,
+                         const rmt_channel_t channel,
+                         const size_t led_count)
+                       : OSThread(),
+                         pin_(output_pin),
+                         channel_(channel),
+                         ledCount_(led_count),
+                         ledBufferSize_(led_count * 3),
+                         sem_(0)
+{
+#if defined(CONFIG_IDF_TARGET_ESP32C3)
+    // ESP32-C3 has four RMT channels with the first two allocated for TX and
+    // second two for RX.
+    HASSERT(channel >= RMT_CHANNEL_0 && channel <= RMT_CHANNEL_1);
+#elif defined(CONFIG_IDF_TARGET_ESP32S2) || defined(CONFIG_IDF_TARGET_ESP32S3)
+    // ESP32-S2 has four RMT channels.
+    // ESP32-S3 has eight RMT channels with the first four allocated for TX and
+    // remaining for RX.
+#elif defined(CONFIG_IDF_TARGET_ESP32)
+    // ESP32 has eight RMT channels.
+    HASSERT(channel >= RMT_CHANNEL_0 && channel <= RMT_CHANNEL_7);
+#endif
+    HASSERT(GPIO_IS_VALID_OUTPUT_GPIO(output_pin));
+}
+
+// destructor
+Esp32WS2812::~Esp32WS2812()
+{
+    free(ledDataBuffer_);
+}
+
+// initialize the perhipheral
+void Esp32WS2812::hw_init()
+{
+    LOG(VERBOSE, "ESP-WS2812(%d): Allocating %u bytes for %zu LEDs", channel_,
+        ledBufferSize_, ledCount_);
+    // Allocate the LED color data buffer
+    ledDataBuffer_ = (uint8_t *)calloc(1, ledBufferSize_);
```

Comment:
how about `malloc(ledBufferSize_)`

Reply:
That would work but would not initialize the memory to zero, a call to `memset` would be needed. Most likely it would compile out to similar code either way.

Reply:
Fixed